### PR TITLE
agent: Discard signals if cannot find sequence ID

### LINKED ISSUE #142: Discard signals if cannot find sequence ID or container is stopped
When a container with terminal enabled is started, it is very likely that it's going to receive a SIGWINCH signal right after it started. But there is an issue here, if the workload exited very quickly, when the signal arrives to the agent, it cannot find the process/seqID (because it has been removed when the process exited).

The race condition happens because at the same time, the agent sends the exit code to the shim, and the shim sends the signal to the agent.

The way to avoid such issues is to proceed the same way it was done with the sdtin coming from IO channel, that is dropping the signal if we have no seqID match. That's why we should discard signals, either because there is no corresponding seqID, or because the container holding those seqID has been stopped. In both cases, we cannot prevent the agent to receive signals, and the best we can do
is to ignore them.

### PR DESCRIPTION
**Pull request by sboeuf**
When a container with terminal enabled is started, it is very likely
that it's going to receive a SIGWINCH signal right after it started.
But there is an issue here, if the workload exited very quickly, when
the signal arrives to the agent, it cannot find the process/seqID.

The race condition happens because at the same time, the agent sends
the exit code to the shim, and the shim sends the signal to the agent.

The way to avoid such issues is to proceed the same way it was done
with the sdtin coming from IO channel, that is dropping the signal
if we have no seqID match. That's why this patch discards signals,
either because there is no corresponding seqID, or because the
container holding those seqID has been stopped. In both cases, we
cannot prevent the agent to receive signals, and the best we can do
is to ignore them.

Fixes #142

### CODE REVIEW

File: agent.go [hunk 1]
```diff
@@ -1392,20 +1400,26 @@ func winsizeCb(pod *pod, data []byte) error {
...
 
-	term, cid, err := pod.findTermFromSeqID(payload.Sequence)
-	if err != nil {
-		return err
+	term, cid := pod.findTermFromSeqID(payload.Sequence)
+	if cid == "" {
+		// The sequence ID could be not found in case the process returned
+		// and the process has been removed from the container map.
+		// We should not error in that case, and discard the received signal.
+		agentLog.Warnf("Could not find sequence ID %d, discard SIGWINCH", payload.Sequence)
```
Comment (jodh-intel) [inline agent.go]: Again, could be useful to have the pod ID logged here.

File: agent.go [hunk 2]
```diff
@@ -1226,11 +1226,19 @@ func killContainerCb(pod *pod, data []byte) error {
...
 	if ctr == nil {
 		return fmt.Errorf("Container %s not found, impossible to signal", payload.ID)
 	}
 
+	status, err := ctr.container.Status()
+	if err != nil {
+		return err
+	}
+	if status == libcontainer.Stopped {
+		agentLog.Info("Container not running anymore, discard signal %s", payload.Signal.String())
```
Comment (jodh-intel) [inline agent.go]: Please can you add the container ID to this message to help with debugging.

File: agent.go [hunk 3]
```diff
@@ -1392,20 +1400,26 @@ func winsizeCb(pod *pod, data []byte) error {
...
 	status, err := ctr.container.Status()
-
 	if err != nil {
 		return err
 	}
 
-	if status != libcontainer.Running {
-		return fmt.Errorf("Container %s not running, impossible to resize window", cid)
+	if status == libcontainer.Stopped {
+		agentLog.Info("Container not running anymore, discard SIGWINCH")
```
Comment (jodh-intel) [inline agent.go]: Pod+Container ID here?

File: agent.go (hunk 2 above)
Comment (grahamwhaley) [reply agent.go]: and if you are there, substitute 'not running' for 'is Stopped' - may as well use the official status state to help us decode the logs later.

**Review by grahamwhaley — COMMENTED** [review]
Good catch.
I'm presuming we saw evidence of the race - maybe just some reported failures in the logs or some error about bad exit codes or something - might be good to note that in the commit or pull message.
With the nits about log messages - l g t m

**Comment by sboeuf** [comment]
@grahamwhaley well I don't have logs from this particular failure since I have discovered this while I was debugging some other issues and by running a simple docker command in a loop, I have realized the issue because I was getting some sporadic failures from the agent. The commit message seems enough verbose to understand what I am fixing here, don't you think ?

**Comment by sboeuf** [comment]
@grahamwhaley @jodh-intel I have addressed your comments.

**Comment by jodh-intel** [comment]
lgtm

[Approved with PullApprove](https://pullapprove.com/clearcontainers/agent/pull-request/143/?utm_source=github-pr&utm_medium=comment-badge&utm_campaign=clearcontainers/agent)

**Review by grahamwhaley — APPROVED** [review]
@sboeuf sure, if we don't have the error logs, np. It's nice practice to include them if we can - it helps folks later on who run into the same issues as they can search more effectively to see if they have been found/fixed before.

lgtm

**Pull request merged**